File: QytetetRuby/lib/controlador_qytetet.rb
#encoding: utf-8
# Francisco Beltrán Sánchez
# Pablo Cordero Romero
require "singleton"
require_relative "qytetet"
require_relative "estado_juego"
require_relative "metodo_salir_carcel"

module ControladorQytetet
    class ControladorQytetet
      include ModeloQytetet
      include Singleton
      
      attr_reader   :modelo
      attr_accessor :nombreJugadores
      
      def initialize
        @modelo = ModeloQytetet::Qytetet.instance
        @nombreJugadores = Array.new
      end
      
      def obtenerOperacionesJuegoValidas
        salida = Array.new
        
        if@modelo.jugadores.empty?
          salida << OpcionMenu.index(:INICIARJUEGO)
        elsif @modelo.estadoJuego == ModeloQytetet::EstadoJuego::JA_PREPARADO
          salida << OpcionMenu.index(:JUGAR)
          salida << OpcionMenu.index(:OBTENERRANKING)
          salida << OpcionMenu.index(:MOSTRARJUGADORACTUAL)
          salida << OpcionMenu.index(:MOSTRARJUGADORES)
          salida << OpcionMenu.index(:MOSTRARTABLERO)
          salida << OpcionMenu.index(:TERMINARJUEGO)
        elsif @modelo.estadoJuego == ModeloQytetet::EstadoJuego::JA_PUEDEGESTIONAR
          salida << OpcionMenu.index(:VENDERPROPIEDAD)
          salida << OpcionMenu.index(:HIPOTECARPROPIEDAD)
          salida << OpcionMenu.index(:CANCELARHIPOTECA)
          salida << OpcionMenu.index(:EDIFICARCASA)
          salida << OpcionMenu.index(:EDIFICARHOTEL)
          salida << OpcionMenu.index(:PASARTURNO)
          salida << OpcionMenu.index(:OBTENERRANKING)
          salida << OpcionMenu.index(:MOSTRARJUGADORACTUAL)
          salida << OpcionMenu.index(:MOSTRARJUGADORES)
          salida << OpcionMenu.index(:MOSTRARTABLERO)
          salida << OpcionMenu.index(:TERMINARJUEGO)
        elsif @modelo.estadoJuego == ModeloQytetet::EstadoJuego::JA_ENCARCELADO
          salida << OpcionMenu.index(:PASARTURNO)
          salida << OpcionMenu.index(:OBTENERRANKING)
          salida << OpcionMenu.index(:MOSTRARJUGADORACTUAL)
          salida << OpcionMenu.index(:MOSTRARJUGADORES)
          salida << OpcionMenu.index(:MOSTRARTABLERO)
          salida << OpcionMenu.index(:TERMINARJUEGO)
        elsif @modelo.estadoJuego == ModeloQytetet::EstadoJuego::JA_CONSORPRESA
          salida << OpcionMenu.index(:APLICARSORPRESA)
          salida << OpcionMenu.index(:OBTENERRANKING)
          salida << OpcionMenu.index(:MOSTRARJUGADORACTUAL)
          salida << OpcionMenu.index(:MOSTRARJUGADORES)
          salida << OpcionMenu.index(:MOSTRARTABLERO)
          salida << OpcionMenu.index(:TERMINARJUEGO)
        elsif @modelo.estadoJuego == ModeloQytetet::EstadoJuego::JA_PUEDECOMPRAROGESTIONAR
          salida << OpcionMenu.index(:VENDERPROPIEDAD)
          salida << OpcionMenu.index(:HIPOTECARPROPIEDAD)
          salida << OpcionMenu.index(:CANCELARHIPOTECA)
          salida << OpcionMenu.index(:EDIFICARCASA)
          salida << OpcionMenu.index(:EDIFICARHOTEL)
          salida << OpcionMenu.index(:COMPRARTITULOPROPIEDAD)
          salida << OpcionMenu.index(:PASARTURNO)
          salida << OpcionMenu.index(:OBTENERRANKING)
          salida << OpcionMenu.index(:MOSTRARJUGADORACTUAL)
          salida << OpcionMenu.index(:MOSTRARJUGADORES)
          salida << OpcionMenu.index(:MOSTRARTABLERO)
          salida << OpcionMenu.index(:TERMINARJUEGO)
        elsif @modelo.estadoJuego == ModeloQytetet::EstadoJuego::JA_ENCARCELADOCONOPCIONDELIBERTAD
          salida << OpcionMenu.index(:INTENTARSALIRCARCELPAGANDOLIBERTAD)
          salida << OpcionMenu.index(:INTENTARSALIRCARCELTIRANDODADO)
          salida << OpcionMenu.index(:OBTENERRANKING)
          salida << OpcionMenu.index(:MOSTRARJUGADORACTUAL)
          salida << OpcionMenu.index(:MOSTRARJUGADORES)
          salida << OpcionMenu.index(:MOSTRARTABLERO)
          salida << OpcionMenu.index(:TERMINARJUEGO)
        elsif @modelo.estadoJuego == ModeloQytetet::EstadoJuego::ALGUNJUGADORENBANCARROTA
          salida << OpcionMenu.index(:OBTENERRANKING)
        end
        
        salida
      end
    
      def necesitaElegirCasilla(opcionMenu)
        salida = false
        
        if( OpcionMenu[opcionMenu] == :HIPOTECARPROPIEDAD || OpcionMenu[opcionMenu] == :CANCELARHIPOTECA || OpcionMenu[opcionMenu] == :EDIFICARCASA || OpcionMenu[opcionMenu] == :EDIFICARHOTEL || OpcionMenu[opcionMenu] == :VENDERPROPIEDAD)
          salida = true
        end  
          
        salida
      end
      
      def obtenerCasillasValidas(opcionMenu)
        salida = Array.new
        
        if OpcionMenu[opcionMenu] == :HIPOTECARPROPIEDAD
          salida = @modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false)
        elsif OpcionMenu[opcionMenu] == :CANCELARHIPOTECA
          salida = @modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(true)
        elsif OpcionMenu[opcionMenu] == :EDIFICARCASA
          salida = @modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false)
        elsif OpcionMenu[opcionMenu] == :EDIFICARHOTEL
          salida = @modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false)
        elsif OpcionMenu[opcionMenu] == :VENDERPROPIEDAD
          salida = @modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false)
        end
          
        salida
      end
      def realizarOperacion(opcionElegida, casillaElegida)
        salida = "\0"
        
        if OpcionMenu[opcionElegida] == :INICIARJUEGO
          @modelo.inicializarJuego(@nombreJugadores)
          salida = "-------------------\nSe ha iniciado el juego\nEmpieza el jugador: #{@modelo.jugadorActual}\n-------------------"
        elsif OpcionMenu[opcionElegida] == :JUGAR
          @modelo.jugar
          salida = "-------------------\nHas tirado el dado y has sacado #{@modelo.getValorDado}\nHas caído en la casilla: #{@modelo.obtenerCasillaJugadorActual}\n-------------------"
        elsif OpcionMenu[opcionElegida] == :APLICARSORPRESA
          salida = "-------------------\nHas aplicado la sorpresa: #{@modelo.cartaActual}\n-------------------"
          @modelo.aplicarSorpresa
        elsif OpcionMenu[opcionElegida] == :INTENTARSALIRCARCELPAGANDOLIBERTAD
          if @modelo.intentarSalirCarcel(ModeloQytetet::MetodoSalirCarcel::PAGANDOLIBERTAD)
            salida = "-------------------\nHas conseguido salir de la carcel :)\n-------------------"
          else
            salida = "-------------------\nNo has conseguido salir de la carcel por falta de saldo :(\n-------------------"
          end
        elsif OpcionMenu[opcionElegida] == :INTENTARSALIRCARCELTIRANDODADO
          if @modelo.intentarSalirCarcel(ModeloQytetet::MetodoSalirCarcel::TIRANDODADO)
              salida = "-------------------\nHas conseguido salir de la carcel :)\n-------------------"
          else
              salida = "-------------------\nNo has obtenido el valor necesario y no has conseguido salir de la carcel :(\n-------------------"
          end
        elsif OpcionMenu[opcionElegida] == :COMPRARTITULOPROPIEDAD
          if @modelo.comprarTituloPropiedad
              salida = "-------------------\nHas comprado la propiedad :)\n-------------------"
          else
              salida = "-------------------\nNo has podido comprar la propiedad por falta de saldo. :(\n-------------------"
          end
        elsif OpcionMenu[opcionElegida] == :HIPOTECARPROPIEDAD
          @modelo.hipotecarPropiedad(casillaElegida)
          salida = "-------------------\nPropiedad hipotecada\n-------------------"
        elsif OpcionMenu[opcionElegida] == :CANCELARHIPOTECA
          if @modelo.cancelarHipoteca(casillaElegida)
            salida = "-------------------\nHipoteca cancelada correctamente\n-------------------"
          else
            salida = "-------------------\nError. No tienes saldo para cancelarla, no eres propietario\n o no esta hipotecada la propiedad\n-------------------"
          end
        elsif OpcionMenu[opcionElegida] == :EDIFICARCASA
          if @modelo.edificarCasa(casillaElegida)
            salida = "-------------------\nCasa edificada correctamente\n-------------------"
          else
            salida = "-------------------\nNo se ha podido edificar la casa\n-------------------"
          end
        elsif OpcionMenu[opcionElegida] == :EDIFICARHOTEL
          if @modelo.edificarHotel(casillaElegida)
            salida = "-------------------\nHotel edificado correctamente\n-------------------"
          else
            salida = "-------------------\nNo se ha podido edificar el hotel\n-------------------"
          end
        elsif OpcionMenu[opcionElegida] == :VENDERPROPIEDAD
          @modelo.venderPropiedad(casillaElegida)
          salida = "-------------------\nPropiedad vendida\n-------------------"
        elsif OpcionMenu[opcionElegida] == :PASARTURNO
          @modelo.siguienteJugador
          salida = "-------------------\nTurno de: #{@modelo.jugadorActual}\n-------------------"
        elsif OpcionMenu[opcionElegida] == :OBTENERRANKING
          @modelo.obtenerRanking
          #salida = "-------------------\n#{@modelo.jugadores}\n-------------------"
        elsif OpcionMenu[opcionElegida] == :MOSTRARJUGADORACTUAL
          salida = "-------------------\n#{@modelo.jugadorActual}\n-------------------"
        elsif OpcionMenu[opcionElegida] == :MOSTRARJUGADORES
          salida = "-------------------\n#{@modelo.jugadores}\n-------------------"
        elsif OpcionMenu[opcionElegida] == :MOSTRARTABLERO
          salida = "-------------------\n#{@modelo.tablero}\n-------------------"
        elsif OpcionMenu[opcionElegida] == :TERMINARJUEGO
          @modelo.obtenerRanking
          salida = "-------------------\nJuego finalizado\n-------------------"
          exit(0)
        end
        
        salida
      end
    end
end

File: QytetetJava/src/controladorqytetet/ControladorQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package controladorqytetet;

import java.util.ArrayList;
import modeloqytetet.EstadoJuego;
import modeloqytetet.MetodoSalirCarcel;
import modeloqytetet.Qytetet;

public class ControladorQytetet {
    private static final ControladorQytetet instance = new ControladorQytetet();
    private static final Qytetet modelo = Qytetet.getInstance();
    private ArrayList<String> nombreJugadores;
    
    private ControladorQytetet(){}
    
    public static ControladorQytetet getInstance(){
        return instance;
    }
    
    public Qytetet getModelo(){
        return modelo;
    }
    
    public ArrayList<String> getNombreJugadores(){
        return nombreJugadores;
    }
    
    public void setNombreJugadores(ArrayList<String> nombres){
        nombreJugadores = nombres;
    }
    
    public ArrayList<Integer> obtenerOperacionesJuegoValidas(){
        ArrayList<Integer> salida = new ArrayList<>();
        
        if(modelo.getJugadores().isEmpty())
                    salida.add(OpcionMenu.INICIARJUEGO.ordinal());
        else
            switch (modelo.getEstadoJuego()) {
                case JA_PREPARADO :  
                    salida.add(OpcionMenu.JUGAR.ordinal());
                    salida.add(OpcionMenu.OBTENERRANKING.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                    salida.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                    salida.add(OpcionMenu.TERMINARJUEGO.ordinal());

                    break;
                case JA_PUEDEGESTIONAR:  
                    salida.add(OpcionMenu.VENDERPROPIEDAD.ordinal());
                    salida.add(OpcionMenu.HIPOTECARPROPIEDAD.ordinal());
                    salida.add(OpcionMenu.CANCELARHIPOTECA.ordinal());
                    salida.add(OpcionMenu.EDIFICARCASA.ordinal());
                    salida.add(OpcionMenu.EDIFICARHOTEL.ordinal());
                    salida.add(OpcionMenu.PASARTURNO.ordinal());
                    salida.add(OpcionMenu.OBTENERRANKING.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                    salida.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                    salida.add(OpcionMenu.TERMINARJUEGO.ordinal());

                    break;
                case JA_ENCARCELADO:  
                    salida.add(OpcionMenu.PASARTURNO.ordinal());
                    salida.add(OpcionMenu.OBTENERRANKING.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                    salida.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                    salida.add(OpcionMenu.TERMINARJUEGO.ordinal());

                    break;
                case JA_CONSORPRESA:
                    salida.add(OpcionMenu.APLICARSORPRESA.ordinal());
                    salida.add(OpcionMenu.OBTENERRANKING.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                    salida.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                    salida.add(OpcionMenu.TERMINARJUEGO.ordinal());

                    break;
                case JA_PUEDECOMPRAROGESTIONAR:
                    salida.add(OpcionMenu.VENDERPROPIEDAD.ordinal());
                    salida.add(OpcionMenu.HIPOTECARPROPIEDAD.ordinal());
                    salida.add(OpcionMenu.CANCELARHIPOTECA.ordinal());
                    salida.add(OpcionMenu.EDIFICARCASA.ordinal());
                    salida.add(OpcionMenu.EDIFICARHOTEL.ordinal());
                    salida.add(OpcionMenu.COMPRARTITULOPROPIEDAD.ordinal());
                    salida.add(OpcionMenu.PASARTURNO.ordinal());
                    salida.add(OpcionMenu.OBTENERRANKING.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                    salida.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                    salida.add(OpcionMenu.TERMINARJUEGO.ordinal());

                    break;
                case JA_ENCARCELADOCONOPCIONDELIBERTAD:
                    salida.add(OpcionMenu.INTENTARSALIRCARCELPAGANDOLIBERTAD.ordinal());
                    salida.add(OpcionMenu.INTENTARSALIRCARCELTIRANDODADO.ordinal());
                    salida.add(OpcionMenu.OBTENERRANKING.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                    salida.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                    salida.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                    salida.add(OpcionMenu.TERMINARJUEGO.ordinal());

                    break;
                case ALGUNJUGADORENBANCARROTA:
                    salida.add(OpcionMenu.OBTENERRANKING.ordinal());

                    break;
                default: 
                    break;
            }
        
        return salida;
    }
    
    public boolean necesitaElegirCasilla(int opcionMenu){
        boolean salida;
        
        switch(OpcionMenu.values()[opcionMenu]){
            case HIPOTECARPROPIEDAD:
            case CANCELARHIPOTECA:
            case EDIFICARCASA:
            case EDIFICARHOTEL:
            case VENDERPROPIEDAD:
                salida = true;
                break;
            default:
                salida = false;
                break;
        }
        
        return salida;
    }
    
    public ArrayList<Integer> obtenerCasillasValidas(int opcionMenu){
        ArrayList<Integer> salida = new ArrayList<>();
        
        switch(OpcionMenu.values()[opcionMenu]){
            case HIPOTECARPROPIEDAD:
                salida = modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false);
                break;
            case CANCELARHIPOTECA:
                salida = modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(true);
                break;
            case EDIFICARCASA:
                salida = modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false);
                break;
            case EDIFICARHOTEL:
                salida = modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false);
                break;
            case VENDERPROPIEDAD:
                salida = modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false);
                break;
            default:
                break;
        }
        
        return salida;
    }
    
    public String realizarOperacion(int opcionElegida, int casillaElegida){
         String salida = "\0";
        
        switch(OpcionMenu.values()[opcionElegida]){
            case INICIARJUEGO:
                modelo.inicializarJuego(nombreJugadores);
                salida = "-------------------"+
                         "\nSe ha iniciado el juego"+
                         "\nEmpieza el jugador: "+modelo.getJugadorActual()+
                         "\n-------------------";
                
                break;
            case JUGAR:
                modelo.jugar();
                salida = "-------------------"+
                         "\nHas tirado el dado y has sacado "+modelo.getValorDado()+
                         "\nHas caído en la casilla: "+modelo.obtenerCasillaJugadorActual()+
                         "\n-------------------";
                
                break;
            case APLICARSORPRESA:
                salida = "-------------------"+
                         "\nHas aplicado la sorpresa: "+modelo.getCartaActual()+
                         "\n-------------------";
                modelo.aplicarSorpresa();
                
                break;
            case INTENTARSALIRCARCELPAGANDOLIBERTAD:
                if(modelo.intentarSalirCarcel(MetodoSalirCarcel.PAGANDOLIBERTAD))   
                    salida = "-------------------"+
                             "\nHas conseguido salir de la carcel :)"+
                             "\n-------------------";
                else
                    salida = "-------------------"+
                             "\nNo has conseguido salir de la carcel por falta de saldo :("+
                             "\n-------------------";
                
                break;
            case INTENTARSALIRCARCELTIRANDODADO:
                if(modelo.intentarSalirCarcel(MetodoSalirCarcel.TIRANDODADO))
                    salida = "-------------------"+
                             "\nHas conseguido salir de la carcel :)"+
                             "\n-------------------";
                else
                    salida = "-------------------"+
                             "\nNo has obtenido el valor necesario y no has conseguido salir de la carcel :("+
                             "\n-------------------";
                
                break;
            case COMPRARTITULOPROPIEDAD:
                if(modelo.comprarTituloPropiedad())
                    salida = "-------------------"+
                             "\nHas comprado la propiedad :)"+
                             "\n-------------------";
                else
                    salida = "-------------------"+
                             "\nNo has podido comprar la propiedad por falta de saldo. :("+
                             "\n-------------------";
                
                break;
            case HIPOTECARPROPIEDAD:
                modelo.hipotecarPropiedad(casillaElegida);
                salida = "-------------------"+
                         "\nPropiedad hipotecada"+
                         "\n-------------------";
                
                break;
            case CANCELARHIPOTECA:
                if(modelo.cancelarHipoteca(casillaElegida))
                    salida = "-------------------"+
                             "\nHipoteca cancelada correctamente"+
                             "\n-------------------";
                else
                    salida = "-------------------"+
                             "\nError. No tienes saldo para cancelarla, no eres propietario "+
                             "\n o no esta hipotecada la propiedad"+
                             "\n-------------------";
                
                break;
            case EDIFICARCASA:
                if(modelo.edificarCasa(casillaElegida))
                    salida = "-------------------"+
                             "\nCasa edificada correctamente"+
                             "\n-------------------";
                else
                    salida = "-------------------"+
                             "\nNo se ha podido edificar la casa"+
                             "\n-------------------";
                
                break;
            case EDIFICARHOTEL:
                if(modelo.edificarHotel(casillaElegida))
                    salida = "-------------------"+
                             "\nHotel edificado correctamente"+
                             "\n-------------------";
                else
                    salida = "-------------------"+
                             "\nNo se ha podido edificar el hotel"+
                             "\n-------------------";
                
                break;
            case VENDERPROPIEDAD:
                modelo.venderPropiedad(casillaElegida);
                salida = "-------------------"+
                         "\nPropiedad vendida"+
                         "\n-------------------";
                break;
            case PASARTURNO:
                modelo.siguienteJugador();
                salida = "-------------------"+
                         "\nTurno de: "+modelo.getJugadorActual()+
                         "\n-------------------";
                
                break;
            case OBTENERRANKING:
                modelo.obtenerRanking();
                break;
            case MOSTRARJUGADORACTUAL:
                salida = "-------------------\n"+
                         modelo.getJugadorActual()+
                         "\n-------------------";
                
                break;
            case MOSTRARJUGADORES:
                salida = "-------------------\n"+
                         modelo.getJugadores()+
                         "\n-------------------";
                
                break;
            case MOSTRARTABLERO:
                salida = "-------------------\n"+
                         modelo.getTablero()+
                         "\n-------------------";
                
                break;
            case TERMINARJUEGO:
                modelo.obtenerRanking();
                salida = "-------------------"+
                         "\nJuego finalizado"+
                         "\n-------------------";
                System.exit(0);
                
                break;
            default:
                break;
        }
        
        return salida;
    }
}
